airflow/providers/sftp/hooks/sftp.py

```diff
@@ -400,3 +406,177 @@ def get_files_by_pattern(self, path, fnmatch_pattern) -> list[str]:
+                )
+                self.known_hosts = "none"
+
+        if host_key is not None:
+            self.known_hosts = f"{conn.host} {host_key}".encode()
+
+    async def _get_conn(self) -> asyncssh.SSHClientConnection:
+        """
+        Asynchronously connect to the SFTP server as an SSH client.
+
+        The following parameters are provided either in the extra json object in
+        the SFTP connection definition
+
+        - key_file
+        - known_hosts
+        - passphrase
+        """
+        conn = await sync_to_async(self.get_connection)(self.sftp_conn_id)
+        if conn.extra is not None:
+            self._parse_extras(conn)
+
+        conn_config = {
+            "host": conn.host,
+            "port": conn.port,
+            "username": conn.login,
+            "password": conn.password,
+        }
+        if self.key_file:
+            conn_config.update(client_keys=self.key_file)
+        if self.known_hosts:
+            if self.known_hosts.lower() == "none":
+                conn_config.update(known_hosts=None)
+            else:
+                conn_config.update(known_hosts=self.known_hosts)
+        if self.private_key:
+            _private_key = asyncssh.import_private_key(self.private_key, self.passphrase)
+            conn_config.update(client_keys=[_private_key])
+        if self.passphrase:
+            conn_config.update(passphrase=self.passphrase)
+        ssh_client_conn = await asyncssh.connect(**conn_config)
+        return ssh_client_conn
+
+    async def list_directory(self, path: str = "") -> list[str] | None:
+        """Returns a list of files on the SFTP server at the provided path."""
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            files = await sftp_client.listdir(path)
+            return sorted(files)
+        except asyncssh.SFTPNoSuchFile:
+            return None
+
+    async def read_directory(self, path: str = "") -> Sequence[asyncssh.sftp.SFTPName] | None:
+        """Returns a list of files along with their attributes on the SFTP server at the provided path."""
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            files = await sftp_client.readdir(path)
+            return files
+        except asyncssh.SFTPNoSuchFile:
+            return None
+
+    async def get_files_and_attrs_by_pattern(
+        self, path: str = "", fnmatch_pattern: str = ""
+    ) -> Sequence[asyncssh.sftp.SFTPName]:
+        """
+        Get the files along with their attributes matching the pattern (e.g. ``*.pdf``) at the provided path.
+
+        if one exists. Otherwise, raises an AirflowException to be handled upstream for deferring
+        """
+        files_list = await self.read_directory(path)
+        if files_list is None:
+            raise FileNotFoundError(f"No files at path {path!r} found...")
```

Comment:
I am thinking on whether we should catch this exception in the Trigger and put then to sleep so that it continues to sense for the availability of this file.

Reply:
+1, sounds reasonable

Reply:
Handled the scenarios here - https://github.com/apache/airflow/pull/37117/commits/82c05921381d6776d89ebeacb64cc7f4e77ed093